python/graphstorm/gconstruct/transform.py

```diff
@@ -1080,23 +1100,55 @@ def process_features(data, ops):
         The data stored as a dict.
     ops : list of FeatTransform
         The operations that transform features.
+    ext_mem: str or None
+        The address of external memory
 
     Returns
     -------
     dict : the key is the data name, the value is the processed data.
     """
     new_data = {}
     for op in ops:
-        res = op(data[op.col_name])
-        assert isinstance(res, dict)
-        for key, val in res.items():
-            # Check if has 1D features. If yes, convert to 2D features
-            if len(val.shape) == 1:
-                if isinstance(val, ExtMemArrayWrapper):
-                    val = val.to_numpy().reshape(-1, 1)
+        hash_hex_feature_path = generate_hash()
+        feature_path = 'feature_{}_{}'.format(op.feat_name, hash_hex_feature_path)
+        feature_path = ext_mem + feature_path if ext_mem is not None \
+            else feature_path
+        if isinstance(op.col_name, str):
+            col_name = [op.col_name]
+        else:
+            col_name = op.col_name
+        tmp_key, wrapper = "", ""
+        for col in col_name:
+            res = op(data[col])
+            assert isinstance(res, dict)
+            for key, val in res.items():
+                # Check if it has 1D features. If yes, convert to 2D features
+                if len(val.shape) == 1:
+                    if isinstance(val, ExtMemArrayWrapper):
+                        val = val.to_numpy().reshape(-1, 1)
+                    else:
+                        val = val.reshape(-1, 1)
+                if len(col_name) == 1:
+                    new_data[key] = val
+                    continue
+                tmp_key = key
+                # Use external memory if it is available
+                if ext_mem is not None:
+                    if not os.path.exists(feature_path):
+                        os.makedirs(feature_path)
+                        wrapper = ExtFeatureWrapper(feature_path, val.shape, val.dtype)
+
+                    hash_hex = generate_hash()
+                    val_path = feature_path + '/{}_{}.npy'.format(col, hash_hex)
```

Comment:
Why we need val_path here?

Reply:
I think we can put the val_path inside the function.